pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,405 @@
+					po := uint8(swarm.Proximity(k.base.Bytes(), peerToRemove.Bytes()))
+					k.knownPeers.Remove(peerToRemove, po)
+				} else {
+					k.logger.Errorf("kademlia manage loop iterator: %v", err)
+				}
+			}
+		}
+	}
+}
+
+// binSaturated indicates whether a certain bin is saturated or not.
+// when a bin is not saturated it means we would like to proactively
+// initiate connections to other peers in the bin.
+func binSaturated(bin, depth uint8, peers *pslice.PSlice) bool {
+	// short circuit for bins which are >= depth
+	if bin >= depth {
+		return false
+	}
+
+	// lets assume for now that the minimum number of peers in a bin
+	// would be 2, under which we would always want to connect to new peers
+	// obviously this should be replaced with a better optimization
+	// the iterator is used here since when we check if a bin is saturated,
+	// the plain number of size of bin might not suffice (for example for squared
+	// gaps measurement)
+
+	size := 0
+	_ = peers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		switch {
+		case po < bin:
+			return true, false, nil
+		case po > bin:
+			return false, true, nil
+		}
+
+		size++
+		return false, false, nil
+	})
+
+	return size >= 2
+}
+
+// recalcDepth returns the kademlia depth. Must be called under lock.
+func (k *Kad) recalcDepth() uint8 {
+	// handle edge case separately
+	if k.connectedPeers.Length() <= nnLowWatermark {
+		return 0
+	}
+	var (
+		peers                        = uint(0)
+		candidate                    = uint8(0)
+		shallowestEmpty, noEmptyBins = k.connectedPeers.ShallowestEmpty()
+	)
+
+	_ = k.connectedPeers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		peers++
+		if peers >= nnLowWatermark {
+			candidate = po
+			return true, false, nil
+		}
+		return false, false, nil
+	})
+
+	if noEmptyBins || shallowestEmpty > candidate {
+		return candidate
+	}
+
+	return shallowestEmpty
+}
+
+// AddPeer adds a peer to the knownPeers list.
+// This does not guarantee that a connection will immediately
+// be made to the peer.
+func (k *Kad) AddPeer(ctx context.Context, addr swarm.Address) error {
+	if k.connectedPeers.Exists(addr) || k.knownPeers.Exists(addr) {
+		return nil
+	}
+
+	po := swarm.Proximity(k.base.Bytes(), addr.Bytes())
+	k.knownPeers.Add(addr, uint8(po))
+
+	select {
+	case k.manageC <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+// Disconnected is called when peer disconnects.
+func (k *Kad) Disconnected(addr swarm.Address) {
+	po := uint8(swarm.Proximity(k.base.Bytes(), addr.Bytes()))
+	k.connectedPeers.Remove(addr, po)
+
+	k.waitNextMu.Lock()
+	k.waitNext[addr.String()] = time.Now().Add(timeToRetry)
+	k.waitNextMu.Unlock()
+
+	k.depthMu.Lock()
+	k.depth = k.recalcDepth()
+	k.depthMu.Unlock()
+	select {
+	case k.manageC <- struct{}{}:
+	default:
+	}
+}
+
+// ClosestPeer returns the closest peer to a given address.
+func (k *Kad) ClosestPeer(addr swarm.Address) (peerAddr swarm.Address, err error) {
```

Comment:
this will be integrated in the 3rd phase, and pricing/accounting estimates will be delegated to the appropriate components through the abstractions that were outlined in the original kademlia presentation